### ISSUE TITLE
Graceful shutdown

### ISSUE THREAD

**Issue by lorenzo** [hasura/graphql-engine#2698]
Currently Hasura, once sent the `SIGINT` signal, terminates all client connections immediately instead of gracefully shutting them down. This means that clients receive a 500 error instead of an actual response.

This is usually a problem as it will look like from the client perspective that the server is having problems, when it may just mean that it is being updated or an instance being replaced.

My suggestion would be to stop responding for new requests, serve all current requests in flight and wait at most 30 seconds for the requests to finish before actually shutting the server down.

If there is any interest in this, I also volunteer to implement it, unless it turns out to be more difficult that I think it is.

**Comment by dsandip** [comment]
@0x777 @lexi-lambda Thoughts?

**Comment by lexi-lambda** [comment]
I think I discussed this with @jberryman very, very briefly when we were in Bangalore. Here’s my thoughts.

Generally, it would be nice to make Hasura shutdown gracefully, waiting until existing connections have finished and ensuring all transactions are committed. However, I don’t know how much effort that would be to implement, and I wonder if it isn’t strictly worth it. Most approaches to scaling services horizontally have this functionality built-in to their gateway/proxy/router, where once an instance has been marked for termination, the gateway will automatically stop routing traffic to it, wait for existing connections to finish, and only then terminate it. I know Heroku’s router does this automatically, and Kubernetes, Docker Swarm, and Amazon ECS all have functionality designed to support it. Generally, the keywords to look for are “connection draining.”

However, as a caveat, my experience with that sort of thing only involves HTTP connections, which are designed to finish quickly. I am not sure what those services do with open websocket connections, nor am I even certain what the appropriate way to handle open websocket connections when shutting down actually is.

tl;dr: I don’t think it makes sense to go to much effort to implement graceful shutdown in Hasura itself if it would just be duplicating functionality present in virtually all gateways, but I am also not certain it would be. @lorenzo, did you have a particular use case in mind that wouldn’t be handled well by draining connections in a gateway server?

**Comment by lorenzo** [comment]
@lexi-lambda currently on mobile so I cannot elaborate on my answer. Just wanted to ask, would you be interested in a pull request for this, provided that the solution is not more difficult than I think it would be?

The appropriate way to close a WebSocket gracefully can only be done with a defined protocol, where a close packet is sent, and the client reconnects immediately, but is n general it is just ok to close the connection whenever. 

I’m more concerned about regular queries and mutations, not about subscriptions since websockets require reconnecting logic anyways

**Comment by lexi-lambda** [comment]
> Just wanted to ask, would you be interested in a pull request for this, provided that the solution is not more difficult than I think it would be?

Certainly. If the change is not too invasive, I’m sure we’d happily accept a PR. :)
